lib/fontist/index.rb

```diff
@@ -0,0 +1,136 @@
+require "forwardable"
+
+module Fontist
+  class Index
+    include Singleton
+
+    class << self
+      extend Forwardable
+      def_delegators :instance, :load_all, :load_formulas, :rebuild
+    end
+
+    def load_all
+      Formulas.register_formulas
+      Registry.formulas
+    end
+
+    def load_formulas(font, style = nil)
+      paths(font, style).map do |path|
+        load_formula(path)
+      end
+    end
+
+    def rebuild
+      @index = {}
+
+      formula_files.each do |file|
+        parse_file(file)
+      end
+
+      save_index
+    end
+
+    private
+
+    def index
+      @index ||= load_index
+    end
+
+    def load_index
+      unless File.exist?(Fontist.formula_index_path)
+        raise Errors::FormulaIndexNotFoundError.new("Please fetch index with `fontist update`.")
+      end
+
+      index = YAML.load_file(Fontist.formula_index_path)
+      index.map do |font, styles|
+        [font.downcase, downcase_keys(styles)]
+      end.to_h
+    end
+
+    def downcase_keys(items)
+      items.map do |key, value|
+        [key.downcase, value]
+      end.to_h
+    end
+
+    def paths(font, style)
+      styles = index[font.downcase]
+      return [] unless styles
+
+      paths = if style
+                styles[style.downcase] || []
+              else
+                styles.values.flatten.uniq
+              end
+
+      paths.map do |file|
+        Fontist.formulas_path.join(file).to_s
+      end
+    end
+
+    def load_formula(path)
+      Formula.new.tap do |f|
+        hash = { "key" => key(path) }.merge(YAML.load_file(path))
+        f.from_hash(hash)
+      end
+    end
```

Comment:
@ronaldtse Am I moving to a right direction? It would load yaml directly to a Formula instance. Will skip generation of classes with FormulaTemplate, FontFormula dsl and Registry.

Reply:
Yes, see https://github.com/fontist/fontist/pull/192#issuecomment-746708324 .